docs/api/net.md

```diff
@@ -63,6 +63,44 @@ Creates a [`ClientRequest`](./client-request.md) instance using the provided
 The `net.request` method would be used to issue both secure and insecure HTTP
 requests according to the specified protocol scheme in the `options` object.
 
+### `net.fetch(input[, init])`
+
+* `input` string | [Request](https://nodejs.org/api/globals.html#request)
+* `init` [RequestInit](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options) (optional)
+
+Returns `Promise<GlobalResponse>` - see [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response).
+
+Sends a request, similarly to how `fetch()` works in the renderer, using
+Chrome's network stack. This differs from Node's `fetch()`, which uses
+Node.js's HTTP stack.
+
+Example:
+
+```js
+async function example () {
+  const response = await net.fetch('https://my.app')
+  if (response.ok) {
+    const body = await response.json()
+    // ... use the result.
+  }
+}
+```
+
+This method will issue requests from the [default
```

Comment:
We have two different session contracts now in the net module with this api, 1) `net.request` that can be configure the session per request on 2) `net.fetch` that only requests on the default session.

* What is the reason to expose `net.fetch`, can it be removed in favor of `session.fetch` ? 
* Should net module in general be a property of session similar to `session.webRequest`, `session.protocol`, `session.cookies` which would allow to not duplicate this api ?

Reply:
> For (2), I think you're suggesting session.defaultSession.net.fetch? And have net become a full submodule of Session?

Yeah, I was suggesting to make net a submodule of session. 

> Given that net is such a small module—just three methods, one of which is a wrapper for another—I'm not sure I feel it's worth it to give an extra component in the namespace (session.defaultSession.net.fetch instead of session.defaultSession.fetch).

We have also `session.netLog` that is another smaller submodule. I think the idea of network related interfaces being under session module encapsulates that the underlying network stack operates on similar boundaries to partition network state. I am not suggesting to block the PR on this discussion item, maybe I separate this out into an issue for disccusion with API wg, thoughts ?

Reply:
I do feel like we should resolve it before we ship it, in order to avoid shipping and then immediately deprecating a method (moving `session.defaultSession.{fetch => net.fetch}`).

I see these disadvantages to the `ses.net.fetch` version:
- the method is a bit less discoverable in the documentation—`fetch` won't be listed under `session`.
- the documentation is confusing. e.g. the [netLog docs](https://www.electronjs.org/docs/latest/api/net-log) don't mention anything about it being a submodule of `session`, or in fact say anything about `session` at all. This could be fixed, but it takes extra effort to explain, vs. the `ses.fetch` version which does not need the concept of a "session submodule" to explain. It's also confusing that `netLog` is _both_ a top-level module _and_ a session submodule.
- `net.isOnline` has nothing to do with any given session

And these advantages:
- the method would be documented in only one place
- we could move the `netLog` functions into `net` (though i would want to keep the old `netLog` name around indefinitely if we did this)
- it structurally suggests to the user that network stuff is always scoped to a session
- it's aesthetically pleasing to me (not a real advantage)

overall i kind of feel like the advantages are heavier here.

Reply:
> net.isOnline has nothing to do with any given session

I think we can move this to `app.isOnline`, it is more of an application level indicator rather than being scoped to session

Reply:
@deepak1556 yeah, though i think in general there's some resistance to throwing stuff on `app`. I personally don't mind it though. However, I _would_ prefer to avoid breaking apps for aesthetic reasons, so if we were to go ahead with moving that API, I'd suggest keeping `net.isOnline` as an alias indefinitely.

Reply:
Still going back and forth on this. I discovered that `ClientRequest` is exported from the `net` module, so it's valid to write:

```
const { net } = require('electron')

const r = new net.ClientRequest(...)
```

I feel like having a _class_ as an instance property on `Session` would be Bad... so should we remove `ClientRequest` as something instantiable? Note that there does appear to be at least some usage of this form, e.g. https://cs.github.com/DracotMolver/soube/blob/8e801be2b8318aab4078790e67aaccb3a402c336/app/assets/js/version/index.js?q=%22new+net.clientRequest%22#L8

Reply:
It also gets quite confusing to write the documentation for `net.fetch` in this case, because the docs don't distinguish between `require('electron').net` and `session.defaultSession.net`.

e.g. here are the docs for `net.fetch` (abridged):

```md
### `net.fetch(input[, init])`

* `input` string | [Request](https://nodejs.org/api/globals.html#request)
* `init` [RequestInit](https://developer.mozilla.org/en-US/docs/Web/API/fetch#options) (optional)

Returns `Promise<GlobalResponse>` - see [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response).

Sends a request, similarly to how `fetch()` works in the renderer, using
Chrome's network stack. This differs from Node's `fetch()`, which uses
Node.js's HTTP stack.
```

How do we describe here that `net.fetch` can be either the global or the session-scoped version?

The best I've got is something like:

> This method can either be called from the global `net` module, or from the [`ses.net`](session.md#sesnet) property on `Session`. In the former case, the request will be issued in the [default session](session.md#sesdefaultsession). In the latter case, the request will be issued in the associated session.

This feels pretty clunky and imprecise to me, but it's the most clear and succinct version I can think of.

I think this maybe tips the scales in favor of `ses.fetch` :/

Reply:
Another documentation kludge: if `net.request` is also callable from the session-scoped `net` module, is it also allowable to override the session through the `session` / `partition` options to `net.request`? If so, that's a bit weird! If not, how do we document that? There's now _three_ different ways to specify the session of a request: (1) `ses.net.request`, (2) `net.request({session: xyz})`, (3) `net.request({partition: 'xyz'})`. What should we do in the case of `ses.net.request({partition: 'xyz'})`?

Reply:
I made an experimental branch here with the changes that would be needed to make it a separate module... I don't really like it though. https://github.com/electron/electron/compare/net-fetch...net-fetch-module?expand=1

I think I'm now on team `ses.fetch`.

Reply:
> It also gets quite confusing to write the documentation for net.fetch in this case, because the docs don't distinguish between require('electron').net and session.defaultSession.net.

How about writing something similar to https://github.com/electron/electron/blob/main/docs/api/protocol.md#using-protocol-with-a-custom-partition-or-session at the top of module ?

Reply:
> Another documentation kludge: if net.request is also callable from the session-scoped net module, is it also allowable to override the session through the session / partition options to net.request? I

Ah this is a concern, we should remove `net.request({session: xyz})` and `net.request({partition: 'xyz'})` if we are to make the `session.net` transition. The former two would just lead to confusing and incorrect usages of the requests API.

Reply:
> I think I'm now on team ses.fetch.

I am swayed to this as well considering the `new net.ClientRequest` and `net.request({session:, partition})` complications.

Reply:
I see that `session.fetch` is atleast on the better path compared to having `net.fetch({session})`. 

Follow-up question, will there be a situation were one cannot use `net.fetch` over `net.request` ? Wondering if we can eventually remove `net.request` in favor of `net.fetch` ?

Reply:
I think `net.request` should remain indefinitely because:

1. It powers `net.fetch` so we need to keep (at least some of) it anyway
2. I think it makes sense to have something that ~matches Node's `http.request`.
3. Existing apps use it, it's nice not to break them.
4. There are some features which don't fit neatly into `fetch`, for instance `getUploadProgress`. Maybe that'll eventually be possible with `fetch`, but it isn't yet. It's easier to jam extra features into `net.request` because it doesn't need to be standards-compliant, and the object/event-based interface is a bit more flexible than the promise-based interface.